vpr/src/route/rr_node.cpp

```diff
@@ -20,9 +21,13 @@ bool t_rr_node::edge_is_configurable(t_edge_size iedge) const {
 
 bool t_rr_node::validate() const {
     //Check internal assumptions about RR node are valid
+    auto& device_ctx = g_vpr_ctx.device();
+
+    const auto& rr_graph = device_ctx.rr_graph;
+
     t_edge_size iedge = 0;
     for (auto edge : edges()) {
-        if (edge < num_configurable_edges()) {
+        if (edge < rr_graph.num_configurable_edges(RRNodeId(id_))) {
```

Comment:
This is inside the ``rr_node_cpp``.  Actually, we are trying to shadow the rr_node.cpp by introducing APIs to rr_graph_view.h
Since this is an internal method. I suggest we rework the API ``validate()`` in rr_graph_view.h/cpp. 
- You move these codes in ``validate()`` from this source file to rr_graph_view.h/cpp. I suggest to add it to rr_graph_view.cpp. The ``validate()`` is not used often. We may not need to set it as inline
- You delete the ``validate()`` function in this source file
- You adapt all the functions that call the ``validate()`` in other source files.

Let me know what you think.
I believe in this process, you can rework all the edge-related APIs. If you need to combine your effort in #1916 , it is o.k. for me.

Reply:
Ok, we can move `validate()` to `rr_graph_view.cpp` and introduce the needed APIs in `rr_graph_view.h`.

Reply:
@tangxifan  
* We can make the changes for `edges()` and `num_edges` this PR and replace the `validate()` to `rr_graph_view.h`
* Or we can keep `edges` and `num_edges()` separate, merge it and then replace the `validate`
Which method is preferable?

Reply:
@behzadmehmood-rs Can you do a grep and see how widely the ``edges()`` and ``num_edges()`` are used in client functions? If they are small (<100 line changes), we can do it in this PR.

Reply:
@tangxifan `validate()` calls `edge_is_configurable` function from `rr_node_cpp` which uses `edge_switch` and both the functions need to be replaced.
* We can implement `edge_switch` in a separate PR and merge in this PR and replace the functions.
* Or we can merge this PR in `master`, implement `edge_switch` in a separate PR, replace the functions and then merge that into master. 
What are your suggestions?

Reply:
@behzadmehmood-rs Can you do a grep on the ``edge_switch``. If code changes are small, we can do it in this PR.

Reply:
@tangxifan changes are not too small for `edge_switch`, more ever after `edge_switch` we will be left with `edge_sink_node` and changes for `edge_switch` + `edge_sink_node` will be much larger than for `edges` + `num_edges`.
Kindly tell how should we move on.

Reply:
Do you have a line count?

Reply:
Around 50 for `edge_switch` and 50 for `edge_sink_node`.

Reply:
Can be more.

Reply:
O.k. It looks significant. We can do it in another PR.

Reply:
So kindly review this PR.

Reply:
@tangxifan  as we are going to replace these functions after `edge_switch()` refactoring, this PR is ready to be reviewed. Kindly review and suggest how should we move forward.

Reply:
Make sense to me.

Reply:
@behzadmehmood-rs I see the runtime overhead. They are mainly from placement and routing. It indicates that our modification slows down the algorithm. I suspect the use of RRGraphView here is the source. Would you mind revert it to previous codes for ``validate()``? See if this is the cause or not.

Reply:
I have run nightly tests after making changes, once the tests are finished, I will compare and upload the results.